repos/system_upgrade/el7toel8/actors/checkpcidrivers/libraries/checkpcidrivers.py

```diff
@@ -0,0 +1,185 @@
+    )
+    unavailable_pci_ids = tuple(
+        pci_id
+        for pci_id in restricted_pci_ids_on_host
+        if restricted_devices_pcis[pci_id].available_rhel8 == 0
+    )
+    unsupported_driver_names = tuple(
+        driver
+        for driver in restricted_driver_names_on_host
+        if restricted_devices_drivers[driver].supported_rhel8 == 0
+    )
+    unsupported_pci_ids = tuple(
+        pci_id
+        for pci_id in restricted_pci_ids_on_host
+        if restricted_devices_pcis[pci_id].supported_rhel8 == 0
+    )
+
+    # Render the report entities
+    title = "Detected PCI drivers that are restricted on RHEL8."
+
+    # Prepare summary
+    summary_partial = "The following drivers are restricted on RHEL 8 system:"
+    # Process unavailable devices
+    if unavailable_driver_names or unavailable_pci_ids:
+        summary_partial += "\n\t Unavailable in RHEL8:"
+        if unavailable_driver_names:
+            summary_partial += "\n\t\t- driver names:\n"
+            summary_partial += "\t\t\t- " + "\n\t\t\t- ".join(
+                unavailable_driver_names
+            )
+        if unavailable_pci_ids:
+            summary_partial += "\n\t\t- pci ids:\n"
+            summary_partial += "\t\t\t- " + "\n\t\t\t- ".join(
+                unavailable_pci_ids
+            )
+
+    # Process unsupported devices
+    if unsupported_driver_names or unsupported_pci_ids:
+        summary_partial += "\n\t Unsupported in RHEL8:"
+        if unsupported_driver_names:
+            summary_partial += "\n\t\t- driver names:\n"
+            summary_partial += "\t\t\t- " + "\n\t\t\t- ".join(
+                unsupported_driver_names
+            )
+        if unsupported_pci_ids:
+            summary_partial += "\n\t\t- pci ids:\n"
+            summary_partial += "\t\t\t- " + "\n\t\t\t- ".join(
+                unsupported_pci_ids
+            )
+
+    reports = [
+        reporting.Title(title),
+        reporting.Summary(summary_partial),
+        reporting.Severity(reporting.Severity.HIGH),
+        reporting.Tags([reporting.Tags.KERNEL, reporting.Tags.DRIVERS]),
+    ]
+    if inhibit_upgrade:
+        reports += [reporting.Flags([reporting.Flags.INHIBITOR])]
+    return reports
+
+
+def create_dict_lookup(driver_names, key):
+    """
+    Creates a dict lookup from the list of Models.
+
+    :param driver_names: List[leapp.models.Model[K, V]]
+    :param key: a key value, that will be used as a primary key to generate a
+        result Dict
+    :return: Dict[leapp.models.Model[K,V].key, leapp.models.Model[K,V]]
+    :raises ValueError: if there are duplicates of leapp.models.Model[K,V].key
+    :raises AttributeError: if key not in K (Model attributes)
+
+    Example:
+    >>> create_dict_lookup([Model(a=1, b=2), Model(a=2, b=3)], key="b") == {
+    >>>     2: Model(a=1, b=2),
+    >>>     3: Model(a=2, b=3),
+    >>> }
+    """
+    lookup = {getattr(item, key): item for item in driver_names}
```

Comment:
I'd check for AttributeError here and raise a ValueError with the corresponding message. Catching AttributeError outside of this helper function is kinda implementation-aware

Reply:
It is documented in the function signature that the function might raise AttributeError (line 96).  I also can't replace the AttribueError with the ValueError, because it is not the same thing. AttributeError will be raised if the key is not in the